app/models/location.rb

```diff
@@ -2,4 +2,53 @@ class Location < ActiveRecord::Base
   belongs_to :person
   belongs_to :district
   belongs_to :state
+
+  after_save :update_nation_builder
+
+  def update_location(address: nil, city: nil, state: nil, zip: nil)
+    if address
+      if district = District.find_by_address(address: address,
+                                             city:    city,
+                                             state:   state,
+                                             zip:     zip)
+        self.address_1 = address
+        self.city      = city
+        self.district  = district
+        self.state     = district.state
+        self.zip_code  = zip if zip = ZipCode.valid_zip_5(zip)
+      else
+        return nil
+      end
+    elsif zip = ZipCode.valid_zip_5(zip) and zip != self.zip_code
```

Comment:
happy to remove redundant self. I learned to include it where leaving it out might make code hard to read (see self.as_json below, which is also redundant).

In this case, `&&` can't be used to replace `and` because it has higher precedence. I can rewrite with parens:
```
(zip = ZipCode.valid_zip_5(zip)) && (zip != self.zip_code)
```
or on two lines if `and` bugs you